DataProcessing/CryptoCoarseFundamentalUniverseDataDownloader.cs

```diff
@@ -0,0 +1,213 @@
+    {
+        private readonly string _market;
+        private readonly string _baseFolder;
+        private readonly string _destinationFolder;
+
+        private Dictionary<Symbol, string> _quoteCurrency = new();
+        private Dictionary<string, Dictionary<Symbol, List<decimal?>>> _dataByDate = new();
+        private Dictionary<Symbol, Security> _existingSecurities = new();
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CryptoCoarseFundamental"/>
+        /// </summary>
+        /// <param name="market">The data vendor market</param>
+        /// <param name="baseFolder">The folder where the base data saved at</param>
+        public CryptoCoarseFundamentalUniverseDataConverter(string market, string baseFolder)
+        {
+            _market = market;
+            _baseFolder = baseFolder;
+            _destinationFolder = Path.Combine(baseFolder, "fundamental", "coarse");
+
+            Directory.CreateDirectory(_destinationFolder);
+        }
+
+        /// <summary>
+        /// Load the local base trade data files, convert into universe file
+        /// </summary>
+        public bool ConvertToUniverseFile()
+        {
+            try
+            {
+                // Get all trade data files in set crypto market data file
+                foreach(var file in Directory.GetFiles(Path.Combine(_baseFolder, "daily"), "*_trade.zip", SearchOption.AllDirectories))
+                {
+                    var dataReader = new LeanDataReader(file);
+                    var baseData = dataReader.Parse();
+
+                    var fileInfo = new FileInfo(file);
+                    LeanData.TryParsePath(fileInfo.FullName, out var symbol, out _, out _);
+
+                    CurrencyPairUtil.DecomposeCurrencyPair(symbol, out var _, out var quoteCurrency);
+                    _quoteCurrency[symbol] = quoteCurrency;
+
+                    _existingSecurities.Add(symbol, CreateSecurity(symbol, quoteCurrency));
+
+                    foreach (TradeBar data in baseData)
+                    {
+                        var dateTime = data.EndTime;
+                        var date = $"{dateTime:yyyyMMdd}";
+                        
+                        if (!_dataByDate.TryGetValue(date, out var content))
+                        {
+                            _dataByDate[date] = content = new Dictionary<Symbol, List<decimal?>>();
+                        }
+                        
+                        content[symbol] = new List<decimal?>{data.Open, data.High, data.Low, data.Close, data.Volume};
+                    }
+                }
+
+                foreach (var date in _dataByDate.Keys.ToList())
+                {
+                    var coarseByDate = _dataByDate[date];
+
+                    foreach (var dataSymbol in coarseByDate.Keys.ToList())
+                    {
+                        _existingSecurities[dataSymbol].SetMarketPrice(new Tick { Value = (decimal)coarseByDate[dataSymbol][^2] });
```

Comment:
- Should avoid the `ToList` calls here, it's iterating twice over the collection?
- Why use `coarseByDate.Keys` if ur later going to access it for the value multiple time bellow `coarseByDate[dataSymbol]`, suggest just using `foreach (var dataSymbol in coarseByDate)` instead
- Should set the security price for all the assets before trying to get conversion rates, else what if the first uses the last security, it's price wouldn't be set?